rfd/0100-proxy-ssh-grpc.md

```diff
@@ -0,0 +1,197 @@
+authors: Tim Ross (tim.ross@goteleport.com)
+state: draft
+---
+
+# RFD 0100 - Use gRPC to proxy SSH connections to Nodes
+
+# Required Approvers
+
+* Engineering @zmb3 && (fspmarshall || espadolini)
+
+## What
+
+Add an alternate transport mechanism to the Proxy for proxying connections
+to Nodes
+
+## Why
+
+One of the primary contributors to `tsh ssh` connection latency is the
+time it takes to perform an SSH handshake. All connections to a Node via
+`tsh` are proxied via a SSH session established with the Proxy. Which means 
+that in order to connect to a Node `tsh` must perform at least two SSH handshakes,
+one with the Proxy to setup the connection transport and another with the
+target Node over the transport to establish the user's SSH connection.
+
+## Details
+
+`tsh ssh` needs to connect to the target Node via the Proxy, but it
+doesn't have to use SSH for that communication. A new gRPC service exposed
+by the Proxy could perform the same operations as the existing SSH server
+but without as much overhead required to establish the session. To minimize
+changes both in `tsh` and on Cluster admins, the existing SSH port can be multiplexed
+to accept both SSH and gRPC by leveraging the TLS ALPN protocol `teleport-proxy-grpc-ssh`.
+Any incoming requests on the SSH listener with said ALPN protocol will be routed
+to the gRPC server and all other requests to the SSH server.
+
+Note: a gRPC server is already exposed via the Proxy web address that users the ALPN protocol
+`teleport-proxy-grpc`. In order to not conflict the new ALPN protocol is required. Reusing the
+existing gRPC server is not an option since it is unauthenticated and has aggressive keep alive
+parameters.
+
+### Proto Definition
+
+The specification is modeled after the [ProxyService](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/legacy/client/proto/proxyservice.proto)
+which is a similar transport mechanism leveraged for Proxy Peering.
+
+```proto
+service ConnectionProxyService {
+  // GetClusterDetails provides cluster information from the perspective of the Proxy
+  rpc GetClusterDetails(GetClusterDetailsRequest) returns (GetClusterDetailsResponse);
+  // DialNode establishes a connection to the target host
+  // 
+  // The bidirectional stream is a transport mechanism to send raw data between
+  // client and server. This allows other protocols(SSH) to be sent through the
+  // tunnel.
+  rpc DialNode(stream DialNodeRequest) returns (stream DialNodeResponse);
+}
+
+// Request for DialNode
+//
+// Connection flow
+// -> Target (client)
+// <- Connection (server)
+// <-> Frame/AgentFrame (client/server)
+// 
+// Separating the Frame and AgentFrame allows
+// SSH and the SSH Agent Protocol to be multiplexed
+// over the same connection.
+message DialNodeRequest {
+  oneof payload {
+    // Contains the information to identify and connect
+    // to the target host
+    Target dial_target = 1;
+    // Raw data to transmit
+    DataFrame frame = 2;
```

Comment:
If this is `DialNode`, could we just have hardcoded `ssh_frame` and `agent_frame` messages instead? Other protocols could just use a different RPC in the future.